### ISSUE TITLE
Allow custom display names for genes

### ISSUE THREAD

**Issue by ValWood** [pombase/canto#2100]
I was really against this idea when it was suggested here:
https://github.com/pombase/canto/issues/2060
It seemed like an unnecessary complication.

However, I would like to do a massive U-turn.
I just did not imagine how many names are not in UniProt (the vast majority). The turnaround for adding these is going to grow the more we suggest.

and, I agree, it is really difficult to know which gene is referred to if you can't make it match the . publication while you are working.

We don't want to add another column, but I think @jseager7 was suggesting that a "display name' could be added in the gene name interface that would over-ride the uniprot name.

However, I did also notice that some entries which do have gene names still  display the identifier see:
<img width="1091" alt="adding_names" src="https://user-images.githubusercontent.com/7359272/68312478-061a0e80-00ab-11ea-9756-6c6fc4372924.png">

**Comment by jseager7** [comment]
> However, I did also notice that some entries which do have gene names still display the identifier see:

That's probably because the fix for this isn't merged into PHI-Canto yet. There's a minor display bug left to fix related to the display of synonyms on the confirmation page (see https://github.com/pombase/canto/issues/1543#issuecomment-549761883). If you would rather not wait for that fix, I can apply the change sooner.

**Comment by jseager7** [comment]
@ValWood do you think we should reopen #2060 now there's been a change of thinking?

**Comment by ValWood** [comment]
I opened this one because the other one had a lot of other content about allele naming which is resolved. I think it's probably clearer to address in a new ticket.

**Comment by kimrutherford** [comment]
> That's probably because the fix for this isn't merged into PHI-Canto yet.

Yep, I think so.  I tried that UniProt ID locally and got the right gene name in Canto.

**Comment by kimrutherford** [comment]
> I think @jseager7 was suggesting that a "display name' could be added in the gene name
> interface that would over-ride the uniprot name.

I think that should be easy-ish to implement code-wise but the user interface for setting the names needs thought.  The gene confirmation page is already complicated.  I there another place in the interface where we could do this name editing?

**Comment by ValWood** [comment]
I think the confirmation page is probably the best place.
We can see if we can simplify?

**Comment by jseager7** [comment]
I agree with Val that this probably belongs on the confirmation page. There's been previous discussion about changing the layout of the page to make the strain selector a bit less confusing. Maybe it's getting to the point that we need to look at the page again.

If we're worried about complicating the page with more widgets, we could just add a generic link somewhere on the page (titled 'Change gene display names' or similar) that sends users to another view &ndash; like a modal dialog (pop-up) or something &ndash; where they can set all the aliases at once. We probably don't need to clutter the page by including a link for each table row.

**Comment by ValWood** [comment]
>If we're worried about complicating the page with more widgets, we could just add a generic link somewhere on the page (titled 'Change gene display names' or similar) that sends users to another view – like a modal dialog (pop-up) or something – where they can set all the aliases at once

I like this idea.  We don't wan to bother  all users with it because somtimes it won't be required. And hopefully we can get some consensus on fungal names and get some added in bulk to the pathogenic fungi when they are unified nomenclature...certainly it will become less of an issue as time goes on.
..and for those who need it, they certainly won't mind the additional hassle because it will make the curation so much easier.

**Comment by kimrutherford** [comment]
Is this needed before going into production?

**Comment by jseager7** [comment]
> Is this needed before going into production?

Based on the fact our curation could end up bottle-necked by UniProt if we don't have a workaround for missing gene names, I'd say this is an important feature. As @ValWood mentioned, I can see how it's going to be a huge pain for community / third-party curators if they're constantly having to mentally cross-reference a UniProt accession identifier with a gene name in a paper. Community curators in particular would probably just give up.

I'm going to add the `for_release` tag, but anyone can remove it if they think otherwise.

**Comment by ValWood** [comment]
We should discuss this in an upcoming meeting, or maybe @jseager7  you could make a proposal?

**Comment by ValWood** [comment]
It adds an extra burden on the curator and the checker trying to figure out which entry is which so it's quite a big time  sink.

**Comment by jseager7** [comment]
@ValWood I made a proposal in a previous comment here: https://github.com/pombase/canto/issues/2100#issuecomment-551015554 about adding a dedicated link to the page that would display a window with controls to change any gene name in the session. I haven't decided exactly how this window would look, but I've created a mockup of the link below (Option 1). The mockup styles the link as a button, but it could just as easily be a text link.

Another possible design (Option 2) is to have an 'Edit…' link as an extra column of each table, that allows editing the gene name only for the respective gene. This might be easier for curators to notice, but it uses up more space, and the meaning is less clear. If we're not worried about space, I could make the meaning more obvious by changing the link text to 'Edit gene names…' or similar.

- - -

### Option 1

![image](https://user-images.githubusercontent.com/37659591/116054738-e3e7c800-a673-11eb-9ad4-4ff719adef69.png)

### Option 2

![image](https://user-images.githubusercontent.com/37659591/116055608-c7985b00-a674-11eb-8778-e136c589535a.png)

**Comment by jseager7** [comment]
Here's what we decided from the discussion today:

* Amend the Gene Edit page to include a dedicated column for the ORF name, and another column for the UniProt gene name (if it exists), which may double as a space for the custom gene name. Alternatively, we could have a third column solely for the custom gene name.

* Add a link to create custom gene names next to whatever column will contain the custom gene names. This was preferred over links at the end of the rows or outside of the tables, and over a single link elsewhere on the page. We think the accepted solution will be the easiest for curators to notice.

* If a custom gene name is added, the old gene name will not be shown anywhere except on the Gene Edit page: all other references to the gene name will be replaced with the custom name.

* Curators should always be allowed to enter a custom name in PHI-Canto, even if a gene name exists in UniProt. In the case where a gene name exists, the custom name will be a synonym; in the case where only an ORF name exists, the custom name is the proposed gene name for submission to UniProt.

* We should periodically submit the gene names suggested by curators to UniProt, so that they can update their records. As described above, these gene names may be synonyms or new candidate names to supersede the ORF name.

* We should include the custom name in the JSON export, so that the PHI-base 5 website has the option to display our chosen gene name in favour of UniProt’s gene name.

* We may need to add a note to the submission page urging curators to alert us when the gene name from UniProt is wrong.

**Comment by jseager7** [comment]
@kimrutherford I think at a minimum, we'll need schema changes to the Curs database to add a column for the custom gene names. We may also need extra columns if we introduce a distinction between ORF names and the primary name provided by UniProt. This could get tricky if the ORF name isn't guaranteed to exist in UniProt (that is, cases where some other database provides the only name for the gene).

**Comment by jseager7** [comment]
I've just checked through all of the UniProtKB entries that we've curated in PHI-Canto, and a number of them don't have a primary name _or_ an ORF name. The only thing that exists is the entry name. See below for a list of links to the entries.

This may affect how we display the data, because in these cases we wouldn't have anything to show in the (proposed) ORF name column either. Currently, Canto defaults to displaying the entry name in the 'Name' column when this happens. I think we need to decide on one of the following options:

* **Option 1**: Use a more generic name for the column that contains the names fetched from UniProt, so it can cover either ORF names or entry names.

* **Option 2**: Add yet another column, specifically for the entry name.

* **Option 3**: Keep the current behaviour for the 'Name' column (where it chooses one name from a hierarchy of names retrieved from UniProt), and add one more column to contain the custom gene name that the curator provides.

@CuzickA, @ValWood, @kimrutherford, do you have any preference on how to handle this?

- - -

### UniProtKB entries with no names

- [A0A2K9YVY7](https://www.uniprot.org/uniprot/A0A2K9YVY7)
- [A0A3B5Y122](https://www.uniprot.org/uniprot/A0A3B5Y122)
- [A0A3B5YY12](https://www.uniprot.org/uniprot/A0A3B5YY12)
- [A0A3B5ZUX1](https://www.uniprot.org/uniprot/A0A3B5ZUX1)
- [A0A3B6HVQ9](https://www.uniprot.org/uniprot/A0A3B6HVQ9)
- [A0A3B6ITE6](https://www.uniprot.org/uniprot/A0A3B6ITE6)
- [A0A3B6JHD1](https://www.uniprot.org/uniprot/A0A3B6JHD1)
- [A0A3B6RF20](https://www.uniprot.org/uniprot/A0A3B6RF20)
- [Q5I6L8](https://www.uniprot.org/uniprot/Q5I6L8)

**Comment by jseager7** [comment]
We may also want to take note of UniProt's documentation on gene names:

https://www.uniprot.org/help/gene_name

**Comment by ValWood** [comment]
> Here's what we decided from the discussion today:

Great summary.

>We may also want to take note of UniProt's documentation on gene names:

Good point, it is good to standardize names across species where possible and UniProt have a good handle on this.  If the pathogen community "roll their own" names they could be ignored longer term in preference of any standardized or unified nomenclature (It might be useful to have a section in the paper about our approach to gene names as this is quite important- pls add to agenda). I managed to convince the pombe community (largely) to be sensible about this and where appropriate follow the naming conventions for well studied orthologous proteins. However, we are curating in an area where there  is much duplication so this is more tricky.

>I've just checked through all of the UniProtKB entries that we've curated in PHI-Canto, and a number of them don't have a primary name or an ORF name

Thanks. I will send this list to Uniprot. This might be to do with a difference between protein names and gene names and what the author decides to apply the name to (In pombe and S. cerevisiae we get around this by having gene and protein names the same except in a protein the first letter is capitalized). Hopefully, UniProt can explain to us.

**Comment by ValWood** [comment]
>Option 1: Use a more generic name for the column that contains the names fetched from UniProt, so it can cover either ORF names or entry names.

I think this would be OK for Canto display. However,  these would need to be stored independently 'behind the scenes' as you will  want to use these different types of names in PHI-base later (a "uniprot entry name" needs to be kept separate from a gene name or other official identifier for this purpose).

**Comment by ValWood** [comment]
Out of curiosity, when UniPRott is updated to name genes, does a Canto 'refresh' replace the of names with  assigned names?
If so, how often does this happen?

**Comment by jseager7** [comment]
> Out of curiosity, when UniProt is updated to name genes, does a Canto 'refresh' replace the of names with assigned names?
If so, how often does this happen?

Canto keeps an internal cache of the gene data that it fetches from UniProt, and while it's possible to empty that cache (and refresh the data from UniProt), it doesn't happen automatically.

In issue https://github.com/pombase/canto/issues/2107, Kim added a feature to the server side of Canto that can be used to clear the gene cache, so this could be configured to run periodically on PHI-Canto's servers; perhaps after every UniProt release.

The only downside of the script is that it causes Canto to be intermittently unresponsive while it's running (not a big issue since the script can be run outside of peak hours).

**Comment by jseager7** [comment]
> However, these [gene names] would need to be stored independently 'behind the scenes' as you will want to use these different types of names in PHI-base later

I think this will be okay because PHI-base 5 should be retrieving its own data from UniProt using the UniProtKB accession number. As long as we have the accession number and the gene name supplied by the curator in the JSON export, we should be able to get all the data we need for the PHI-base display.

**Comment by ValWood** [comment]
OK, we should find out when UniProt updates happen, and do immediately after this.

**Comment by jseager7** [comment]
UniProt recommends subscribing to their RSS feed to stay notified about updates, but the feed includes other news, not just UniProt releases. I couldn't get any RSS application working with the feed, so I'll just have to remember to check the page every month or so.

https://www.uniprot.org/help/update_notification
